libbzip2-rs-sys/src/bzlib.rs

```diff
@@ -689,43 +703,33 @@ macro_rules! ADD_CHAR_TO_BLOCK {
     };
 }
 
-unsafe fn copy_input_until_stop(strm: &mut bz_stream, s: &mut EState) -> bool {
+fn copy_input_until_stop(strm: &mut bz_stream, s: &mut EState) -> bool {
     let mut progress_in = false;
 
     match s.mode {
         Mode::Running => loop {
             if s.nblock >= s.nblockMAX {
                 break;
             }
-            if strm.avail_in == 0 {
+            if let Some(b) = strm.read_byte() {
+                progress_in = true;
+                ADD_CHAR_TO_BLOCK!(s, b as u32);
+            } else {
                 break;
             }
-            progress_in = true;
-            ADD_CHAR_TO_BLOCK!(s, *(strm.next_in as *mut u8) as u32);
-            strm.next_in = (strm.next_in).offset(1);
-            strm.avail_in = (strm.avail_in).wrapping_sub(1);
-            strm.total_in_lo32 = (strm.total_in_lo32).wrapping_add(1);
-            if strm.total_in_lo32 == 0 {
-                strm.total_in_hi32 = (strm.total_in_hi32).wrapping_add(1);
-            }
         },
         Mode::Idle | Mode::Flushing | Mode::Finishing => loop {
             if s.nblock >= s.nblockMAX {
                 break;
             }
-            if strm.avail_in == 0 {
-                break;
-            }
```

Comment:
can't you keep this check here to be sure that the behavior is the same? we then know that the if let below will always succeed, so we're doing the check twice, but I'm not convinced the behavior is truly equivalent with the checks swapped.

Reply:
In this case both checks do a break and the checks themself don't have any side-effects, so swapping them is guaranteed to have the same effect.